### PR TITLE
feat: add a new API to prove multi continuous blocks in one circuit

### PR DESCRIPTION
**Pull request by lispc**
Done.

@HAOYUatHZ @Lawliet-Chan the new multi block proving api will be like this. You can have a preview.

Suggest use this condition "min(0.5M gas, 10 blocks)" for grouping blocks.

### CODE REVIEW

File: zkevm/tests/integration.rs [hunk 1]
```diff
@@ -122,9 +121,25 @@ fn test_mock_prove_all_with_circuit<C: TargetCircuit>(
...
+    use zkevm::circuit::EvmCircuit;
+
+    init();
+    let mut block_results = Vec::new();
+    for block_number in 1..=15 {
+        let trace_path = format!("tests/traces/bridge/{:02}.json", block_number);
+        let block_result = get_block_result_from_file(trace_path);
+        block_results.push(block_result);
+    }
+    Prover::mock_prove_target_circuit_packing::<EvmCircuit>(&block_results, true).unwrap();
```
Comment (lispc) [inline zkevm/tests/integration.rs:136]: @noel2004  can you add zktrie and poseidon here and make sure they pass?

Comment (noel2004) [reply zkevm/tests/integration.rs:136]: ok

Comment (noel2004) [reply zkevm/tests/integration.rs:136]: Test has passed in `test_mock_prove_all_target_circuits_packing`:

(I add logs for rows being used)

>[2022-08-29T05:14:40Z INFO  zkevm::prover] start mock prove zktrie
>[2022-08-29T05:14:40Z INFO  zkevm::circuit] zktrie use rows 1766
>[2022-08-29T05:14:42Z INFO  zkevm::prover] mock prove zktrie done
>[2022-08-29T05:14:42Z INFO  zkevm::prover] start mock prove poseidon
>[2022-08-29T05:14:42Z INFO  zkevm::circuit] poseidon use rows 33250
>[2022-08-29T05:14:51Z INFO  zkevm::prover] mock prove poseidon done

File: zkevm/src/circuit/builder.rs [hunk 2]
```diff
@@ -49,29 +49,37 @@ fn extend_address_to_h256(src: &Address) -> [u8; 32] {
     bts.as_slice().try_into().expect("32 bytes")
 }
 
-pub fn block_result_to_witness_block<F: Field>(
+pub fn block_result_to_witness_block(
     block_result: &BlockResult,
 ) -> Result<Block<Fr>, anyhow::Error> {
-    let chain_id = if let Some(tx_trace) = block_result.block_trace.transactions.get(0) {
+    block_results_to_witness_block(&[block_result.clone()])
```
Comment (Thegaram) [inline zkevm/src/circuit/builder.rs:55]: Could also use `slice::from_ref` here.

```suggestion
    block_results_to_witness_block(std::slice::from_ref(block_result))
```

File: zkevm/src/circuit/builder.rs [hunk 3]
```diff
@@ -95,125 +103,107 @@ pub fn decode_bytecode(bytecode: &str) -> Result<Vec<u8>, anyhow::Error> {
...
     cdb.insert(decode_bytecode(EMPTY_ACCOUNT_CODE)?);
 
-    for execution_result in &block.execution_results {
-        if let Some(bytecode) = execution_result.byte_code.clone() {
-            cdb.insert(decode_bytecode(&bytecode)?);
-        }
-    }
+    for block in blocks {
+        for execution_result in &block.execution_results {
+            if let Some(bytecode) = execution_result.byte_code.clone() {
```
Comment (Thegaram) [inline zkevm/src/circuit/builder.rs:117]:
```suggestion
            if let Some(bytecode) = &execution_result.byte_code
```

File: zkevm/src/prover.rs [hunk 4]
```diff
@@ -314,17 +336,25 @@ impl Prover {
...
+        let (circuit, instance) = C::from_block_results(block_results)?;
         let mut transcript = PoseidonWrite::<_, _, Challenge255<_>>::init(vec![]);
 
         let instance_slice = instance.iter().map(|x| &x[..]).collect::<Vec<_>>();
 
         let public_inputs: &[&[&[Fr]]] = &[&instance_slice[..]];
 
         info!(
-            "Create {} proof of block {}",
+            "Create {} proof of block {}, batch len {}",
```
Comment (HAOYUatHZ) [inline zkevm/src/prover.rs:354]:
```suggestion
            "Create {} proof of block {} ... block {}, batch len {}",
```

File: zkevm/src/prover.rs [hunk 5]
```diff
@@ -314,17 +336,25 @@ impl Prover {
...
             C::name(),
-            block_result.block_trace.hash
+            block_results[0].block_trace.hash,
```
Comment (HAOYUatHZ) [inline zkevm/src/prover.rs:356]:
```suggestion
            block_results[0].block_trace.hash,
            block_results[len-1].block_trace.hash,
```

File: zkevm/src/prover.rs [hunk 6]
```diff
@@ -349,7 +379,7 @@ impl Prover {
         info!(
             "Create {} proof of block {} Successfully!",
```
Comment (HAOYUatHZ) [inline zkevm/src/prover.rs:380]:
```suggestion
            "Create {} proof of block {} ... block {} Successfully!",
```

File: zkevm/src/prover.rs [hunk 7]
```diff
@@ -349,7 +379,7 @@ impl Prover {
...
             C::name(),
-            block_result.block_trace.hash
+            block_results[0].block_trace.hash
```
Comment (HAOYUatHZ) [inline zkevm/src/prover.rs:382]:
```suggestion
            block_results[0].block_trace.hash
            block_results[len-1].block_trace.hash
```

File: zkevm/src/circuit/builder.rs [hunk 8]
```diff
@@ -95,125 +103,107 @@ pub fn decode_bytecode(bytecode: &str) -> Result<Vec<u8>, anyhow::Error> {
...
-                    _ => {}
+            for (k, val) in s_map {
+                let mut k_buf: [u8; 32] = [0; 32];
+                k.to_big_endian(&mut k_buf[..]);
+                let val_proof: mpt::StorageProof = val.as_slice().try_into()?;
+                let val = verify_proof_leaf(val_proof, &k_buf);
+
+                if val.key.is_some() {
+                    // a valid leaf
+                    acc.storage.insert(*k, *val.data.as_ref());
```
Comment (HAOYUatHZ) [inline zkevm/src/circuit/builder.rs:201]: seem I need to dig into zkevmi-circuits codes if I want to undertstand more deeply.... will it overwrite if using same `k`?

Comment (lispc) [reply zkevm/src/circuit/builder.rs:201]: yes. it is by design. zkevm circuits needs the storage before a block and replay all storage write/read in memory. So here we traverse all storage io with reverse order

**Comment by lispc** [comment]
@noel2004 @HAOYUatHZ can you review and merge this PR?

**Review by HAOYUatHZ — APPROVED** [review]

**Pull request merged**